src/main/java/net/minecraftforge/common/capabilities/ICapabilityProvider.java

```diff
@@ -6,30 +6,61 @@
 
 import net.minecraft.core.Direction;
-import net.minecraftforge.common.util.LazyOptional;
+
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
 public interface ICapabilityProvider
 {
     /**
-     * Retrieves the Optional handler for the capability requested on the specific side.
-     * The return value <strong>CAN</strong> be the same for multiple faces.
-     * Modders are encouraged to cache this value, using the listener capabilities of the Optional to
-     * be notified if the requested capability get lost.
+     * Retrieves the {@link Capability} for the capability requested on the specific side.<br>
+     * The return value may be the same for multiple faces.<br>
+     * Callers should cache the returned instance and not make repeated calls to this method when possible.<br>
+     * Contractually, unless invalidated, the returned instance will always be the same for the same arguments.<br>
      *
-     * @param cap The capability to check
-     * @param side The Side to check from,
-     *   <strong>CAN BE NULL</strong>. Null is defined to represent 'internal' or 'self'
+     * @param type The {@link CapabilityType} to check for.
+     * @param direction The {@link Direction} to check from, which may be null.  Null represents "internal".<br>
+     * If this provider is not direction-sensitive, only null should be passed.
      * @return The requested an optional holding the requested capability.
+     * 
+     * @see {@link #isDirectionSensitive()}
+     */
+    @NotNull <T> Capability<T> getCapability(@NotNull final CapabilityType<T> type, final @Nullable Direction direction);
+
+    /**
+     * Utility for {@link #getCapability(Capability, Direction)} that calls with a null direction.<br>
+     * Do not override this method - only override the contextual variant!
+     * 
+     * @see {@link #getCapability(Capability, Direction)}
      */
-    @NotNull <T> LazyOptional<T> getCapability(@NotNull final Capability<T> cap, final @Nullable Direction side);
+    @NotNull default <T> Capability<T> getCapability(@NotNull final CapabilityType<T> type) {
+        return getCapability(type, null);
+    }
 
-    /*
-     * Purely added as a bouncer to sided version, to make modders stop complaining about calling with a null value.
-     * This should never be OVERRIDDEN, modders should only ever implement the sided version.
+    /**
+     * Checks if this capability provider is direction-sensitive.<br>
+     * The return of this method should be static, i.e. a provider should always or never be direction-sensitive.<br>
+     * Of the default providers, {@link ItemStack} is not direction sensitive.
+     * @return If directions have importance in {@link #getCapability(Capability, Direction)}
      */
-    @NotNull default <T> LazyOptional<T> getCapability(@NotNull final Capability<T> cap) {
-        return getCapability(cap, null);
+    default boolean isDirectionSensitive()
```

Comment:
I don't think this is useful, you know that if you're attaching to `ItemStack`s the direction is useless in most cases. And there might be some use cases where a specific cap wants direction-awareness inside of an item. (e.g. something like an AE2 storage cell _could_ in theory be direction-aware if the ME Drive directly forwards the insertion request with the side context).

Reply:
Technically this is overridable by TE's and Entities, which are free to notify the universe that they do not have special capability returns on a side-wise basis.  It might end up just being annoying, however.

Reply:
That sounds like it will cause issues at some point, e.g. maybe I want to attach a direction-aware capability to all subclasses of `BaseContainerBlockEntity`, however some mod might just randomly decide to override that method and that would break my attachment. Imo the system should try to be as broad as possible instead of attempting to limit use cases.

Reply:
Yeah, probably just going to drop it based on lack of utility.